### ISSUE TITLE
Define and code enhanced implementation of Google recommendation for Angular project file structure

### ISSUE THREAD

**Issue by kosz** [kosz/generator-modular#2]
> Note: this content of this text has been drastically modified based on the comments provided.
> Note: it's expected that this document and it's links to other generator-modular wiki pages, to change and be improved with time.
## Rationale

As I'm coding [generator-modular](https://github.com/kosz/generator-modular), I am trying to provide a scalable and robust standard for File Structure organization. This approach will obviously be based on the [Google Recommendation for Angular project file structure](https://docs.google.com/document/d/1XXMvReO8-Awi1EZXAXS4PzDzdNvV6pGcuaF4Q9821Es/mobilebasic?pli=1), however, when trying to anticipate potential problems with large & growing generator-modular based projects, I believe I may have found a few issues, which I would like to discuss here openly, and have them validated and approved, or constructively invalidated, by the community. 
## A few issues with the Google Recommendation
- the way google recommendation sees the project root
- file naming conventions using mixed dash and snake case
- the lack of clarity regarding route modularity 
- the mixed components directory 
- ambiguity regarding how modules should be used

This list is short and not very descriptive. For a more in depth analysis of these thoughts please have a look at this wiki page : https://github.com/kosz/generator-modular/wiki/Issues-with-the-Google-Best-Practice-Recommendation-for-Angular-App-Structure
## Side by Side example

Here is a side by side example to illustrate the difference between the Google Recommendation and the Modular Proposal. Make sure to click on the image to zoom in.
![Modular Recommendation](https://raw.githubusercontent.com/kosz/angular-webapp-demo-app/master/modular.png)

> > > Camel case in file names violates the Google Recommendation. This is a mistake I made when creating this image ( due to using this convention in recent projects ), and will be fixed in later revisions  
## Conclusion

So basically the Modular proposal, indicates that base level API objects such as routes,filters,directives or services should live in a type based directory structure, but encapsulate all their files and other components. 

That means that we do not drop ALL the directives in the directives folder, we only drop the base level, reusable directives. If a directive contains other directives which only are used inside it, then the child directives should live inside the folder of the parent directive.

**Comment by dancancro** [comment]
I don't think I'm understanding this.  As I read the suggestion from Google (the indentation of the doc needs some help) it looks like under those rules you can have sub things inside of main things, which seems to be what you're wanting to provide.  They call them "sub-sections". For example, sub-section 1-1 is inside of subsection 1.  So I'm confused about what's lacking.

Also, I can't tell where in the Google proposal or your proposal the login-card directive is supposed to go.  Could you add that in there?  Also, maybe add "filters/" and "services/" to your proposed diagram and show where the login service and login related filters would go as well as such things that are not exclusively tied to the login module.

Also this 

> if we have a route called login, and a login-card directive which will ONLY be used in the login route, then the login-card directive will be encapsulated within the login route.

seems perfectly in accord with the Google proposal.  Just locate the login-card directive somewhere in or under the login directory.

You might want to make a heading in this doc labeled "Problems" or something and then include a succinct (one short sentence per.  give it a good think. brevity is genius) enumeration of the problems you see with the Google proposal.  It's a little strange that while such a list is missing you do have a large heading for "The Confusion".  I like the "Angular Recommendation..." and "Proposed Modular..." parts though.

Good luck, and kudos for wanting to make the world better!

**Comment by kosz** [comment]
@dancancro thanks a lot, your comments have helped me improve my thoughts and the text describing them on github a lot better. I have modified the text of this issue above, based on some of your suggestions, and started working on a more indepth [comparison page](https://github.com/kosz/generator-modular/wiki/Issues-with-the-Google-Best-Practice-Recommendation-for-Angular-App-Structure),  which although it can paint a decent picture for now, still needs some work. 

I will update this idea more in the next few days, and once i feel it's clearer to grasp by the readers ( or someone shuts it down completely ) I will start looking for additional input from as many sources as possible. 

The goal with this github ticket for me, is knowing that I'm building this yeoman generator as robustly as possible. While I would love to stick to the Google Recommendation 100%, I feel it falls short, when trying to anticipate issues in larger projects, which is what I'm trying to do, for generator-modular.
